### PR TITLE
Check in && Check out

### PR DESCRIPTION
**Pull request by joaochioda**

### CODE REVIEW

File: api/src/main/java/xyz/luan/dextra/parking/models/rentHistory/RentsAction.java [hunk 1]
```diff
@@ -0,0 +1,88 @@
...
+import java.util.logging.Logger;
+
+
+public class RentsAction extends Action<Rents> {
+
+	private static final Logger LOGGER = Logger.getLogger(RentsAction.class.getName());
+	private MatrixBatchBuilder matrixBuilder;
+
+	private void matrixColumn() {
+		int coluna = 0;
```
Comment (luanpotter) [inline api/src/main/java/xyz/luan/dextra/parking/models/rentHistory/RentsAction.java]: column? consistency with the function name

File: api/src/main/java/xyz/luan/dextra/parking/models/rentHistory/RentsAction.java [hunk 2]
```diff
@@ -0,0 +1,88 @@
...
+		this.matrixBuilder.setValueZeroBased(0, coluna++, "Carro");
+		this.matrixBuilder.setValueZeroBased(0, coluna++, "Data Check In");
+		this.matrixBuilder.setValueZeroBased(0, coluna++, "Hora Check In");
+		this.matrixBuilder.setValueZeroBased(0, coluna++, "Data Check Out");
+		this.matrixBuilder.setValueZeroBased(0, coluna, "Hora Check Out");
+	}
+
+
+	private void matrixFill() {
+		int linha = 1;
```
Comment (luanpotter) [inline api/src/main/java/xyz/luan/dextra/parking/models/rentHistory/RentsAction.java]: line

File: api/src/main/java/xyz/luan/dextra/parking/models/rentHistory/RentsAction.java [hunk 3]
```diff
@@ -0,0 +1,88 @@
...
+		List<Rents> rentHistories = yawp(Rents.class).list();
+		rentHistories.sort(Comparator.comparing(Rents::getCheckInDate));
+
+		LOGGER.info("Found " + rentHistories.size() + " rentHistory.");
+
+		for (Rents rents : rentHistories) {
+			this.matrixBuilder.setValueZeroBased(linha, 0, rents.getPersonEmail());
+			Car car = yawp(Car.class).where("id","=",rents.getCarId()).only();
```
Comment (luanpotter) [inline api/src/main/java/xyz/luan/dextra/parking/models/rentHistory/RentsAction.java]: `("id", "=", rents.getCarId())` (missing spaces)

File: api/src/main/java/xyz/luan/dextra/parking/models/rentHistory/RentsAction.java [hunk 4]
```diff
@@ -0,0 +1,88 @@
...
+			this.matrixBuilder.setValueZeroBased(linha, 1, car.getName());
+			Timestamp timestampCheckIn = new Timestamp(rents.getCheckInDate());
+			this.matrixBuilder.setValueZeroBased(linha, 2, new SimpleDateFormat("dd/MM/yyyy").format(timestampCheckIn));
+			this.matrixBuilder.setValueZeroBased(linha, 3, new SimpleDateFormat("HH:mm:ss").format(timestampCheckIn));
+			Timestamp timestampCheckOut = new Timestamp(rents.getCheckOutDate());
+			this.matrixBuilder.setValueZeroBased(linha, 4, (rents.getCheckOutDate() == 0 ? "0" : new SimpleDateFormat("dd/MM/yyyy").format(timestampCheckOut)));
+			this.matrixBuilder.setValueZeroBased(linha, 5, (rents.getCheckOutDate() == 0 ? "0" : new SimpleDateFormat("HH:mm:ss").format(timestampCheckOut)));
```
Comment (luanpotter) [inline api/src/main/java/xyz/luan/dextra/parking/models/rentHistory/RentsAction.java]: how about a
```
formatAsDate(long timestamp) {
    return timestamp == 0 ? "0" : new SimpleDateFormat("dd/MM/yyyy").format(new Timestamp(timestamp));
}
```
and similar a `formatAsTime` to be used twice each above?

File: web/src/Components/Home/CheckOutModal.jsx [hunk 5]
```diff
@@ -0,0 +1,89 @@
...
+            <ModalFooter>
+                <Button onClick={() => toggleModal()}>Cancelar</Button>{' '}
+                <Button
+                    disabled={!modalSelectedCar}
+                    onClick={() => {
+                        if (isCheckIn) {
+                            checkIn(modalSelectedCar.id)
+                        } else {
+                            checkOut(modalSelectedCar.id)
+                        }
```
Comment (luanpotter) [inline web/src/Components/Home/CheckOutModal.jsx]: `(isCheckIn ? checkIn : checkOut)(modalSelectedCar.id)`

File: web/src/Components/Home/Home.jsx [hunk 6]
```diff
@@ -64,10 +65,9 @@ class Home extends Component {
     this.ids.push(store.on('login', data => {
       this.setState({
         personEmail: data.email,
-      }
-      );
+      });
+      this.getRentedCars(data.email)
```
Comment (luanpotter) [inline web/src/Components/Home/Home.jsx]: missing semicolon on line 69

File: web/src/Components/Home/Home.jsx [hunk 7]
```diff
@@ -24,9 +24,10 @@ class Home extends Component {
   constructor(props) {
     super(props);
     this.state = {
-      cars: [],
-      rentModalVisibility: false,
-      returnModalVisibility: false,
+      cars: null,
+      isCheckIn: false,
+      checkInModalVisibility: false,
+      checkInOutModalVisibility: false,
```
Comment (luanpotter) [inline web/src/Components/Home/Home.jsx]: `checkIn` and `checkInOut`? did you mean `checkIn` and `checkOut`? feels like `checkInOut = in + out`

File: web/src/Components/Home/Home.jsx [hunk 8]
```diff
@@ -170,125 +156,74 @@ class Home extends Component {
...
-  renderReturnModal(rentedCars) {
-    const { returnModalVisibility, returnModalSelectedCar } = this.state;
+  renderCheckInModal() {
+    const { checkInModalVisibility, selectedCar } = this.state;
 
     return (
-      <Modal isOpen={returnModalVisibility} toggle={() => this.toggleReturnModal()}>
-        <ModalHeader toggle={() => this.toggleReturnModal()}>Devolver carro</ModalHeader>
+      <Modal isOpen={checkInModalVisibility} toggle={() => this.toggleCheckInModal()}>
+        <ModalHeader toggle={() => this.toggleCheckInModal()}>{`Alugar carro`}</ModalHeader>
```
Comment (luanpotter) [inline web/src/Components/Home/Home.jsx]: simply `Alugar carro` (no need for the braces and the backticks)

File: web/src/Components/Home/Home.jsx [hunk 9]
```diff
@@ -170,125 +156,74 @@ class Home extends Component {
...
       </Modal>
-    );
+    )
+  }
+  
+  getRentedCars(personEmail) {
+      yawp('/rents').list().then(rentedCars => {
+        const filteredRentedCars = rentedCars.filter(rent => rent.personEmail === personEmail && rent.checkOutDate === 0);
+        this.setState({ rentedCars: filteredRentedCars })
+       }).catch(ex => console.error('Falha ao trazer carros. ', ex))
```
Comment (luanpotter) [inline web/src/Components/Home/Home.jsx]: `carros. '` -> extra space here

File: web/src/Components/Home/Home.jsx [hunk 10]
```diff
@@ -170,125 +156,74 @@ class Home extends Component {
...
   }
 
   render() {
-    const { rentModalVisibility, cars, personEmail } = this.state;
-    const rentedCars = cars.filter(car => car.currentPersonEmail === personEmail && !car.isAvailable);
+    const { rentedCars, checkInModalVisibility, checkInOutModalVisibility, cars, isCheckIn } = this.state;
 
+    if (!rentedCars || !cars) {
+      return('');
```
Comment (luanpotter) [inline web/src/Components/Home/Home.jsx]: return is not a function!!!!!!
simply:
```
    return '';
````

File: web/src/Components/Home/Home.jsx [hunk 11]
```diff
@@ -297,24 +232,38 @@ class Home extends Component {
...
         />
         <button
-          disabled={rentedCars.length <= 0}
           type="button"
-          className="btn floating-button bottom-floating-button"
-          onClick={() => this.toggleReturnModal()}
+          className="btn floating-button bottom-floating-button right"
+          onClick={() => {
+            this.setState({ isCheckIn: true })
+            this.toggleCheckInOutModal()
```
Comment (luanpotter) [inline web/src/Components/Home/Home.jsx]: missing semicolons

File: web/src/Components/Home/Home.jsx [hunk 12]
```diff
@@ -297,24 +232,38 @@ class Home extends Component {
...
         </button>
-        {this.renderRentModal()}
-        {this.renderReturnModal(rentedCars)}
+
+        <button
+          type="button"
+          className="btn floating-button bottom-floating-button left"
+          onClick={() => {
+            this.setState({ isCheckIn: false })
+            this.toggleCheckInOutModal()
```
Comment (luanpotter) [inline web/src/Components/Home/Home.jsx]: missing semicolons

File: web/src/Components/Home/Home.jsx (hunk 7 above)
Comment (joaochioda) [reply web/src/Components/Home/Home.jsx]: yeah checkInOut is a modal that share the same code with checkIn and checkOut, there is another modal checkIn that is different, that's why we have similar names

File: web/src/Components/Home/Home.jsx [hunk 13]
```diff
@@ -170,125 +156,74 @@ class Home extends Component {
...
-  renderReturnModal(rentedCars) {
-    const { returnModalVisibility, returnModalSelectedCar } = this.state;
+  renderCheckInModal() {
+    const { checkInModalVisibility, selectedCar } = this.state;
 
     return (
-      <Modal isOpen={returnModalVisibility} toggle={() => this.toggleReturnModal()}>
-        <ModalHeader toggle={() => this.toggleReturnModal()}>Devolver carro</ModalHeader>
+      <Modal isOpen={checkInModalVisibility} toggle={() => this.toggleCheckInModal()}>
+        <ModalHeader toggle={() => this.toggleCheckInModal()}>Alugar carro</ModalHeader>
```
Comment (eliasreis54) [inline web/src/Components/Home/Home.jsx]: Eu acho que aqui pode só mandar a função.... Não precisa colocar uma função anônima

File: web/src/Components/Home/Home.jsx [hunk 14]
```diff
@@ -170,125 +156,74 @@ class Home extends Component {
...
         </ModalBody>
         <ModalFooter>
-          <Button onClick={() => this.toggleReturnModal()}>Cancelar</Button>{' '}
-          <Button
-            disabled={isObjectEmpty(returnModalSelectedCar)}
-            onClick={() => this.returnCar()}
-          >
-            Confirmar
-          </Button>
+          <Button color="secondary" onClick={() => this.toggleCheckInModal()}>Cancelar</Button>{' '}
```
Comment (eliasreis54) [inline web/src/Components/Home/Home.jsx]: O mesmo...

File: web/src/Components/Home/Home.jsx (hunk 13 above)
Comment (almeidadelucas) [reply web/src/Components/Home/Home.jsx]: Feito

File: web/src/Components/Home/Home.jsx (hunk 14 above)
Comment (almeidadelucas) [reply web/src/Components/Home/Home.jsx]: Feito

**Pull request merged**